### PR TITLE
Add trait formatting

### PR DESCRIPTION
**Pull request by ConnorGBrewster**
Fixes issue #78. 
This PR includes formatting for traits along with the const and type trait/impl items.

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -607,6 +607,112 @@ pub fn format_struct(context: &RewriteContext,
...
+           item.node {
+        let mut result = String::new();
+        let header = format!("{}{}trait {}",
+                             format_visibility(item.vis),
+                             format_unsafety(unsafety),
+                             item.ident);
+
+        result.push_str(&header);
+
+        // TODO: Add max_width checking
```
Comment (nrc) [inline src/items.rs]: You should do something with this - either uncomment and use it or remove or convert to a FIXME with an issue and some explanation

File: src/items.rs [hunk 2]
```diff
@@ -607,6 +607,112 @@ pub fn format_struct(context: &RewriteContext,
...
+                                                     offset + result.len(),
+                                                     mk_sp(item.span.lo, body_lo)));
+        result.push_str(&generics_str);
+
+        let trait_bound_str = try_opt!(rewrite_trait_bounds(context,
+                                                            type_param_bounds,
+                                                            offset,
+                                                            context.config.max_width));
+
+        if offset.width() + result.len() + trait_bound_str.len() > context.config.max_width {
```
Comment (nrc) [inline src/items.rs]: Could you add a comment here and elsewhere explaining where we are in the output text and why you might break here. (Also elsewhere in this function).

File: src/items.rs [hunk 3]
```diff
@@ -607,6 +607,112 @@ pub fn format_struct(context: &RewriteContext,
...
+                                                             &generics.where_clause,
+                                                             context.config,
+                                                             context.config.item_brace_style,
+                                                             context.block_indent,
+                                                             where_budget,
+                                                             context.config.where_density,
+                                                             "{",
+                                                             context.config.where_trailing_comma,
+                                                             None));
+        if !where_clause_str.contains('\n') &&
```
Comment (nrc) [inline src/items.rs]: where clause should always be on a newline unless we can fit the whole trait on one line. For that you'll also need to test that the body is empty and that result does not contain `\n`, as well as the tests you have so far.

You can leave this for later and always insert a newline before the where clause if you like.

File: src/items.rs [hunk 4]
```diff
@@ -914,6 +1020,75 @@ pub fn rewrite_static(prefix: &str,
     rewrite_assign_rhs(context, lhs, expr, remaining_width, context.block_indent).map(|s| s + ";")
 }
 
+pub fn rewrite_associated_type(prefix: &str,
```
Comment (nrc) [inline src/items.rs]: prefix is always `type`, I think, so you don't need the param

File: src/visitor.rs [hunk 5]
```diff
@@ -348,11 +362,24 @@ impl<'a> FmtVisitor<'a> {
                               ii.span,
                               ii.id);
             }
-            ast::ImplItemKind::Const(..) => {
-                // FIXME: Implement
-            }
-            ast::ImplItemKind::Type(_) => {
-                // FIXME: Implement
+            ast::ImplItemKind::Const(ref ty, ref expr) => {
+                let rewrite = rewrite_static("const",
```
Comment (nrc) [inline src/visitor.rs]: should this be `rewrite_associated_static`? Could you unify the two functions?

**Comment by nrc** [comment]
Awesome work, thanks!

There's a few nits inline and I think this needs a few more tests, in particular to test that we do line break where we ought to when we get to the max width.

File: src/items.rs (hunk 1 above)
Comment (ConnorGBrewster) [inline src/items.rs]: Looks like it can be removed, I did add max width checking. Looks like I forgot to remove that comment.

**Comment by ConnorGBrewster** [comment]
@nrc I made the changes you suggest, let me know if I need to fix anything else.

**Comment by nrc** [comment]
Thank you!
